src/osd/OSD.cc

```diff
@@ -9151,18 +9151,18 @@ void OSD::do_recovery(
       pg->discover_all_missing(*rctx.query_map);
       if (rctx.query_map->empty()) {
 	string action;
-        if (pg->state_test(PG_STATE_BACKFILL)) {
+        if (pg->state_test(PG_STATE_BACKFILLING)) {
 	  auto evt = PG::CephPeeringEvtRef(new PG::CephPeeringEvt(
 	    queued,
 	    queued,
-	    PG::CancelBackfill()));
+	    PG::DeferBackfill(cct->_conf->osd_recovery_retry_interval)));
```

Comment:
I agree it's not ideal. I'm adding a todo item for a RECOVERY_STUCK state, but I don't want to pull it into this PR because i suspect it will take some work to get the state changes right.. and this is causing real pain now.

Reply:
Is there a ticket for this?